File: static/classroom.js
/* Classroom dashboard front-end. */

const $ = (sel, root = document) => root.querySelector(sel);
const $$ = (sel, root = document) => Array.from(root.querySelectorAll(sel));

const fmtPct = (v, digits = 1) => (v == null ? "—" : `${(v * 100).toFixed(digits)}%`);
const fmtBrier = (v) => (v == null ? "—" : Number(v).toFixed(3));
const fmtInt = (v) => (v == null ? "—" : String(v));
const fmtDate = (iso) => {
  if (!iso) return "—";
  try { return iso.slice(0, 19).replace("T", " "); } catch { return iso; }
};
const esc = (s) => {
  if (s == null) return "";
  return String(s).replace(/[&<>"']/g, (c) => ({
    "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;",
  }[c]));
};

async function fetchJSON(url) {
  const r = await fetch(url);
  if (!r.ok) throw new Error(`${url}: ${r.status}`);
  return r.json();
}

function dirClass(d) {
  if (d === "up") return "dir-up";
  if (d === "down") return "dir-down";
  if (d === "flat") return "dir-flat";
  return "";
}

// ---- Overview ----

async function loadOverview() {
  const o = await fetchJSON("/api/classroom/overview");
  $("#overview-hdr").textContent = `${o.n_students} students · ${o.n_open_predictions} open · ${o.n_resolved} resolved`;
  const cards = [
    { label: "Students", value: o.n_students, sub: "in the roster" },
    { label: "Predictions today", value: o.today_predictions, sub: o.today_date },
    { label: "Open predictions", value: o.n_open_predictions, sub: "awaiting horizon" },
    { label: "Resolved", value: o.n_resolved, sub: `${o.n_correct} correct` },
    { label: "Classroom hit rate", value: fmtPct(o.classroom_hit_rate), sub: "across all resolutions" },
    { label: "Overall Brier", value: fmtBrier(o.overall_brier), sub: "weighted by resolutions" },
  ];
  $("#overview-cards").innerHTML = cards.map((c) => `
    <div class="cls-card">
      <div class="label">${esc(c.label)}</div>
      <div class="value">${esc(c.value)}</div>
      <div class="sub">${esc(c.sub)}</div>
    </div>
  `).join("");
  $("#cls-last").textContent = o.recent_days.length
    ? `${o.recent_days.length} day(s) with predictions`
    : "no daily firings yet";
}

// ---- Leaderboard ----

let leaderboardCache = null;
let activeTab = "resolved";

let projectionByName = {};

function renderLeaderboard() {
  if (!leaderboardCache) return;
  const rows = activeTab === "resolved" ? leaderboardCache.students : leaderboardCache.unresolved;
  const headers = `
    <div class="cls-row leaderboard header">
      <div>#</div><div>Student</div><div>Technique</div>
      <div class="right">Resolved</div><div class="right">Hit rate</div>
      <div class="right">Brier-30</div><div class="right">Lifetime Brier</div>
      <div class="right">Projected</div><div class="right">Retired</div>
    </div>
  `;
  const body = rows.map((r, i) => {
    const proj = projectionByName[r.name] || {};
    const projected = proj.projected_brier;
    const retired = proj.n_setups_retired || 0;
    const trajCls = (proj.trajectory_delta != null && proj.trajectory_delta > 0.02)
      ? "dir-up" : (proj.trajectory_delta != null && proj.trajectory_delta < -0.02)
      ? "dir-down" : "";
    return `
    <div class="cls-row leaderboard">
      <div class="muted-small">${i + 1}</div>
      <div class="name"><a href="/classroom/student/${encodeURIComponent(r.name)}">${esc(r.name)}</a></div>
      <div><span class="technique-pill">${esc(r.technique || "—")}</span></div>
      <div class="right">${fmtInt(r.total_resolved)}</div>
      <div class="right">${fmtPct(r.hit_rate)}</div>
      <div class="right">${fmtBrier(r.rolling_brier_30)}</div>
      <div class="right">${fmtBrier(r.lifetime_brier)}</div>
      <div class="right ${trajCls}">${fmtBrier(projected)}</div>
      <div class="right">${retired > 0 ? `<span class="status-resolved-wrong">${retired}</span>` : "0"}</div>
    </div>
  `;}).join("");
  $("#leaderboard").innerHTML = rows.length ? headers + body : `<div class="muted small">No students in this bucket yet.</div>`;
  $("#leaderboard-summary").textContent = `${rows.length} student${rows.length === 1 ? "" : "s"}`;
}

async function loadLeaderboard() {
  leaderboardCache = await fetchJSON("/api/classroom/leaderboard?limit=150");
  renderLeaderboard();
}

$$(".cls-tab").forEach((btn) => {
  btn.addEventListener("click", () => {
    $$(".cls-tab").forEach((b) => b.classList.remove("active"));
    btn.classList.add("active");
    activeTab = btn.dataset.tab;
    renderLeaderboard();
  });
});

// ---- By technique ----

async function loadTechniqueTable() {
  const { techniques } = await fetchJSON("/api/classroom/by_technique");
  const headers = `
    <div class="cls-row tech header">
      <div>#</div><div>Technique</div>
      <div class="right">Students</div><div class="right">Resolved</div>
      <div class="right">Hit rate</div><div class="right">Avg Brier</div>
      <div>Sample students</div>
    </div>
  `;
  const body = techniques.map((t, i) => `
    <div class="cls-row tech">
      <div class="muted-small">${i + 1}</div>
      <div class="name">${esc(t.technique)}</div>
      <div class="right">${fmtInt(t.n_students)}</div>
      <div class="right">${fmtInt(t.total_resolved)}</div>
      <div class="right">${fmtPct(t.hit_rate)}</div>
      <div class="right">${fmtBrier(t.avg_lifetime_brier)}</div>
      <div class="muted-small">${t.students.slice(0, 3).map((s) => esc(s)).join(", ")}${t.students.length > 3 ? ` +${t.students.length - 3}` : ""}</div>
    </div>
  `).join("");
  $("#technique-table").innerHTML = headers + body;
}

// ---- Hall of Fame / Wall of Shame ----

function renderHall(entries, kind) {
  if (!entries.length) {
    return `<div class="muted small">No ${kind} entries yet. They're written when a high-confidence (≥85%) prediction resolves.</div>`;
  }
  return entries.map((e) => {
    const first2lines = e.markdown.split("\n").slice(0, 2).join(" — ").replace(/^#\s+/, "");
    return `
      <div class="cls-hall-entry">
        <h3>${esc(first2lines)}</h3>
        <div class="small">${esc(e.slug)}</div>
      </div>
    `;
  }).join("");
}

async function loadHall() {
  const [hof, wos] = await Promise.all([
    fetchJSON("/api/classroom/hall_of_fame"),
    fetchJSON("/api/classroom/wall_of_shame"),
  ]);
  $("#hof-list").innerHTML = renderHall(hof.entries, "Hall of Fame");
  $("#wos-list").innerHTML = renderHall(wos.entries, "Wall of Shame");
}

// ---- Usage ----

async function loadUsage() {
  const u = await fetchJSON("/api/classroom/usage");
  if (!u.exists || u.total_calls === 0) {
    $("#usage-content").innerHTML = `<div class="muted small">No llm_persona calls logged yet. Once today's run fires, this panel will show day/persona breakdowns.</div>`;
    return;
  }
  const daysHeader = `
    <div class="cls-row usage-day header">
      <div>Date</div><div class="right">Calls</div><div class="right">Skipped</div>
      <div class="right">Errors</div><div class="right">Avg ms</div>
    </div>
  `;
  const days = u.by_day.map((d) => `
    <div class="cls-row usage-day">
      <div>${esc(d.date)}</div>
      <div class="right">${fmtInt(d.calls)}</div>
      <div class="right">${fmtInt(d.skipped)}</div>
      <div class="right">${d.errors > 0 ? `<span class="status-resolved-wrong">${esc(d.errors)}</span>` : "0"}</div>
      <div class="right">${fmtInt(d.avg_ms)}</div>
    </div>
  `).join("");

  const personasHeader = `
    <div class="cls-row usage-persona header">
      <div>Persona</div><div class="right">Calls</div><div class="right">Errors</div><div class="right">Avg ms</div>
    </div>
  `;
  const personas = u.by_persona.map((p) => `
    <div class="cls-row usage-persona">
      <div>${esc(p.persona)}</div>
      <div class="right">${fmtInt(p.calls)}</div>
      <div class="right">${p.errors > 0 ? `<span class="status-resolved-wrong">${esc(p.errors)}</span>` : "0"}</div>
      <div class="right">${fmtInt(p.avg_ms)}</div>
    </div>
  `).join("");

  $("#usage-content").innerHTML = `
    <div class="usage-section">
      <div>
        <h3>By day</h3>
        <div class="cls-table">${daysHeader}${days}</div>
      </div>
      <div>
        <h3>By persona</h3>
        <div class="cls-table">${personasHeader}${personas}</div>
      </div>
    </div>
  `;
}

// ---- Latest predictions ----

async function loadLatest() {
  const { predictions } = await fetchJSON("/api/classroom/latest_predictions?limit=80");
  $("#latest-count").textContent = `${predictions.length} most recent`;
  const headers = `
    <div class="cls-row log header">
      <div>When</div><div>Student</div><div>Ticker</div><div>Dir</div>
      <div class="right">H</div><div class="right">Conf</div><div>Setup</div><div>Reasoning</div>
    </div>
  `;
  const body = predictions.map((p) => `
    <div class="cls-row log">
      <div class="muted-small">${esc(fmtDate(p.created_at))}</div>
      <div><a href="/classroom/student/${encodeURIComponent(p.student)}">${esc(p.student)}</a></div>
      <div>${esc(p.ticker)}</div>
      <div class="${dirClass(p.direction)}">${esc(p.direction)}</div>
      <div class="right">${esc(p.time_horizon)}</div>
      <div class="right">${esc(p.confidence)}<span class="confidence-bar"><i style="width:${esc(p.confidence)}%"></i></span></div>
      <div class="muted-small">${esc(p.setup_type || "—")}</div>
      <div class="muted-small">${esc((p.reasoning || "").slice(0, 140))}${(p.reasoning || "").length > 140 ? "…" : ""}</div>
    </div>
  `).join("");
  $("#latest-list").innerHTML = headers + body;
}

function fmtDelta(v) {
  if (v == null) return "—";
  const sign = v > 0 ? "+" : "";
  return `${sign}${Number(v).toFixed(3)}`;
}

function renderMiniStudentList(rows, options = {}) {
  if (!rows || rows.length === 0) {
    return `<div class="muted small">No students in this bucket.</div>`;
  }
  const showField = options.showField || "trajectory_delta";
  const fieldLabel = options.fieldLabel || "Δ";
  const headers = `
    <div class="cls-row tech header">
      <div>#</div><div>Student</div>
      <div class="right">Brier</div><div class="right">Projected</div>
      <div class="right">${esc(fieldLabel)}</div><div class="right">Setups</div>
      <div>Tech</div>
    </div>
  `;
  const body = rows.map((r, i) => `
    <div class="cls-row tech">
      <div class="muted-small">${i + 1}</div>
      <div class="name"><a href="/classroom/student/${encodeURIComponent(r.name)}">${esc(r.name)}</a></div>
      <div class="right">${fmtBrier(r.observed_brier)}</div>
      <div class="right">${fmtBrier(r.projected_brier)}</div>
      <div class="right ${r[showField] > 0 ? "dir-up" : r[showField] < 0 ? "dir-down" : ""}">${fmtDelta(r[showField])}</div>
      <div class="right">${esc(r.n_setups_active_recent || 0)}/${esc(r.n_setups || 0)} <span class="muted-small">(${esc(r.n_setups_retired || 0)} retired)</span></div>
      <div class="muted-small">${esc(r.technique)}</div>
    </div>
  `).join("");
  return headers + body;
}

async function loadLearningLab() {
  let lab;
  try { lab = await fetchJSON("/api/classroom/learning_lab"); } catch { return; }
  if (!lab.exists) {
    $("#learning-lab").hidden = false;
    $("#learning-lab-overview").innerHTML = `<div class="muted small">No projection data yet. Run <code>python scripts/project_scores.py</code> to compute projections (also runs in the nightly cycle).</div>`;
    return;
  }
  $("#learning-lab").hidden = false;
  const c = lab.classroom || {};
  $("#learning-lab-stamp").textContent = lab.generated_at ? `as of ${lab.generated_at.slice(0, 19).replace("T", " ")}` : "";
  const cards = [
    { label: "Students with projection", value: c.n_active_recent, sub: `of ${c.n_students}` },
    { label: "Avg observed Brier", value: fmtBrier(c.avg_observed_brier), sub: "across all resolutions" },
    { label: "Avg projected Brier", value: fmtBrier(c.avg_projected_brier), sub: "going forward" },
    { label: "Avg learning gain", value: fmtDelta(c.avg_learning_gain), sub: "vs naive 70% baseline" },
    { label: "Retired setups", value: c.n_retired_setups, sub: "cumulative" },
  ];
  $("#learning-lab-overview").innerHTML = cards.map((cd) => `
    <div class="cls-card">
      <div class="label">${esc(cd.label)}</div>
      <div class="value">${esc(cd.value)}</div>
      <div class="sub">${esc(cd.sub)}</div>
    </div>
  `).join("");

  $("#lab-improving").innerHTML = renderMiniStudentList(lab.improving,
    { showField: "trajectory_delta", fieldLabel: "Δ obs→proj" });
  $("#lab-deteriorating").innerHTML = renderMiniStudentList(lab.deteriorating,
    { showField: "trajectory_delta", fieldLabel: "Δ obs→proj" });
  $("#lab-most-learned").innerHTML = renderMiniStudentList(lab.most_learned,
    { showField: "learning_gain", fieldLabel: "Δ naive→proj" });
  $("#lab-retirement").innerHTML = renderMiniStudentList(lab.retirement_candidates,
    { showField: "trajectory_delta", fieldLabel: "Δ obs→proj" });
}

async function loadProjections() {
  try {
    const data = await fetchJSON("/api/classroom/projections");
    if (data && data.students) {
      projectionByName = Object.fromEntries(data.students.map((s) => [s.name, s]));
    }
  } catch (e) { console.error(e); }
}

async function init() {
  try { await loadOverview(); } catch (e) { console.error(e); }
  try { await loadProjections(); } catch (e) { console.error(e); }
  try { await loadLearningLab(); } catch (e) { console.error(e); }
  try { await loadLeaderboard(); } catch (e) { console.error(e); }
  try { await loadTechniqueTable(); } catch (e) { console.error(e); }
  try { await loadHall(); } catch (e) { console.error(e); }
  try { await loadUsage(); } catch (e) { console.error(e); }
  try { await loadLatest(); } catch (e) { console.error(e); }
}

init();

File: app/classroom.py
"""Classroom data layer for the dashboard.

Reads from the `classroom` project under the projects root.
Read-only — never writes to the classroom directory.
"""
from __future__ import annotations

import json
import math
from collections import defaultdict
from datetime import datetime, timezone
from pathlib import Path

from . import projects as proj

CLASSROOM_ROOT = proj.PROJECTS_ROOT / "classroom"
STUDENTS_DIR = CLASSROOM_ROOT / "students"
DAILY_DIR = CLASSROOM_ROOT / "data" / "daily"
USAGE_LOG = CLASSROOM_ROOT / "data" / "usage_log.jsonl"
HOF_DIR = CLASSROOM_ROOT / "vault" / "Hall of Fame"
WOS_DIR = CLASSROOM_ROOT / "vault" / "Wall of Shame"
ROSTER_PATH = CLASSROOM_ROOT / "config" / "students.yml"
PROJECTIONS_PATH = CLASSROOM_ROOT / "data" / "projections.json"


def _safe_name(name: str) -> bool:
    """Sanity-check student name to keep filesystem reads safe."""
    return bool(name) and all(c.isalnum() or c in "-_" for c in name) and ".." not in name


def _read_score(path: Path) -> dict | None:
    try:
        return json.loads(path.read_text())
    except (OSError, json.JSONDecodeError):
        return None


def _read_jsonl(path: Path, last_n: int | None = None) -> list[dict]:
    if not path.exists():
        return []
    try:
        lines = path.read_text(encoding="utf-8").splitlines()
    except OSError:
        return []
    if last_n is not None and len(lines) > last_n:
        lines = lines[-last_n:]
    out = []
    for line in lines:
        line = line.strip()
        if not line:
            continue
        try:
            out.append(json.loads(line))
        except json.JSONDecodeError:
            continue
    return out


def _score_sortkey(s: dict) -> float:
    """Lower is better — Brier preferred; fall back to negative hit rate so
    high-hit-rate-low-Brier rises to the top together."""
    b = s.get("rolling_brier_30")
    if b is not None:
        return float(b)
    hr = s.get("hit_rate")
    if hr is not None:
        return 1.0 - float(hr)  # convert to "Brier-ish" loss
    return 0.9  # unresolved students sink toward the bottom


def _roster_names() -> set[str]:
    """Names of students in the current config/students.yml (so we filter
    out orphans left over from earlier param-grid configurations)."""
    if not ROSTER_PATH.exists():
        return set()
    try:
        import yaml  # FastAPI app already depends on PyYAML transitively
        data = yaml.safe_load(ROSTER_PATH.read_text()) or {}
        return {s["name"] for s in (data.get("students") or []) if "name" in s}
    except Exception:
        return set()


def list_students() -> list[dict]:
    out = []
    if not STUDENTS_DIR.exists():
        return out
    live = _roster_names()
    for d in sorted(STUDENTS_DIR.iterdir()):
        if not d.is_dir():
            continue
        if live and d.name not in live:
            continue
        score = _read_score(d / "score.json")
        log_path = d / "log.jsonl"
        n_open = 0
        n_total = 0
        last_pred_at = None
        if log_path.exists():
            for p in _read_jsonl(log_path):
                n_total += 1
                if p.get("status") == "open":
                    n_open += 1
                last_pred_at = p.get("created_at") or last_pred_at
        out.append({
            "name": d.name,
            "technique": (score or {}).get("technique"),
            "params": (score or {}).get("technique_params"),
            "total_predictions": n_total,
            "open_predictions": n_open,
            "total_resolved": (score or {}).get("total_resolved", 0),
            "total_correct": (score or {}).get("total_correct", 0),
            "hit_rate": (score or {}).get("hit_rate"),
            "rolling_brier_30": (score or {}).get("rolling_brier_30"),
            "lifetime_brier": (score or {}).get("lifetime_brier"),
            "current_streak": (score or {}).get("current_streak", 0),
            "best_streak": (score or {}).get("best_streak", 0),
            "calibration_shift": (score or {}).get("calibration_shift", 0),
            "last_prediction_at": last_pred_at,
        })
    return out


def leaderboard(limit: int = 50) -> list[dict]:
    rows = list_students()
    # Only rank students with at least one resolved prediction; the rest sit
    # in an unresolved bucket on the page.
    ranked = [r for r in rows if (r.get("total_resolved") or 0) > 0]
    ranked.sort(key=_score_sortkey)
    return ranked[:limit]


def unresolved_students() -> list[dict]:
    rows = list_students()
    return [r for r in rows if (r.get("total_resolved") or 0) == 0]


def by_technique() -> list[dict]:
    rows = list_students()
    agg = defaultdict(lambda: {"n": 0, "resolved": 0, "correct": 0, "brier_sum": 0.0, "brier_n": 0,
                                "open": 0, "students": []})
    for r in rows:
        t = r.get("technique") or "unknown"
        a = agg[t]
        a["n"] += 1
        a["resolved"] += r.get("total_resolved", 0)
        a["correct"] += r.get("total_correct", 0)
        a["open"] += r.get("open_predictions", 0)
        if r.get("lifetime_brier") is not None:
            a["brier_sum"] += float(r["lifetime_brier"])
            a["brier_n"] += 1
        a["students"].append(r["name"])
    out = []
    for t, a in agg.items():
        hit_rate = (a["correct"] / a["resolved"]) if a["resolved"] else None
        avg_brier = (a["brier_sum"] / a["brier_n"]) if a["brier_n"] else None
        out.append({
            "technique": t,
            "n_students": a["n"],
            "total_resolved": a["resolved"],
            "total_correct": a["correct"],
            "total_open": a["open"],
            "hit_rate": hit_rate,
            "avg_lifetime_brier": avg_brier,
            "students": a["students"],
        })
    out.sort(key=lambda x: (x["avg_lifetime_brier"] is None, x["avg_lifetime_brier"] or 1.0))
    return out


def student_detail(name: str) -> dict | None:
    if not _safe_name(name):
        return None
    base = STUDENTS_DIR / name
    if not base.exists():
        return None
    score = _read_score(base / "score.json") or {}
    log = _read_jsonl(base / "log.jsonl", last_n=200)
    profile_path = base / "profile.md"
    profile_md = ""
    if profile_path.exists():
        try:
            profile_md = profile_path.read_text(encoding="utf-8")[:20_000]
        except OSError:
            pass
    # Sort log: most recent first
    log.sort(key=lambda p: p.get("created_at", ""), reverse=True)
    return {
        "name": name,
        "score": score,
        "profile_md": profile_md,
        "log": log,
    }


def hall_entries(kind: str, limit: int = 50) -> list[dict]:
    """List Hall-of-Fame or Wall-of-Shame markdown entries."""
    target = HOF_DIR if kind == "fame" else WOS_DIR
    if not target.exists():
        return []
    out = []
    for path in sorted(target.glob("*.md"), reverse=True)[:limit]:
        try:
            text = path.read_text(encoding="utf-8")
        except OSError:
            continue
        out.append({
            "slug": path.stem,
            "filename": path.name,
            "markdown": text[:15_000],
        })
    return out


def usage_summary(days: int = 14) -> dict:
    if not USAGE_LOG.exists():
        return {"by_day": [], "by_persona": [], "total_calls": 0, "exists": False}
    records = _read_jsonl(USAGE_LOG)
    if not records:
        return {"by_day": [], "by_persona": [], "total_calls": 0, "exists": True}
    cutoff_date = (datetime.now(timezone.utc).date()).isoformat()
    # Window: last N days inclusive of today
    from datetime import timedelta
    cutoff = (datetime.now(timezone.utc) - timedelta(days=days)).strftime("%Y-%m-%d")
    by_day = defaultdict(lambda: {"calls": 0, "skipped": 0, "errors": 0, "duration_ms": 0})
    by_persona = defaultdict(lambda: {"calls": 0, "errors": 0, "duration_ms": 0})
    total_calls = 0
    for r in records:
        day = r.get("at", "")[:10]
        if day < cutoff:
            continue
        kind = r.get("kind")
        if kind == "llm_persona_call":
            by_day[day]["calls"] += 1
            by_day[day]["duration_ms"] += int(r.get("duration_ms", 0))
            persona = r.get("persona", "?")
            by_persona[persona]["calls"] += 1
            by_persona[persona]["duration_ms"] += int(r.get("duration_ms", 0))
            if int(r.get("exit_code", 0)) != 0:
                by_day[day]["errors"] += 1
                by_persona[persona]["errors"] += 1
            total_calls += 1
        elif kind == "llm_persona_skipped":
            by_day[day]["skipped"] += 1
    out_days = []
    for day in sorted(by_day):
        s = by_day[day]
        avg_ms = (s["duration_ms"] // s["calls"]) if s["calls"] else 0
        out_days.append({"date": day, "calls": s["calls"], "skipped": s["skipped"],
                          "errors": s["errors"], "avg_ms": avg_ms})
    out_personas = []
    for persona in sorted(by_persona, key=lambda p: -by_persona[p]["calls"]):
        s = by_persona[persona]
        avg_ms = (s["duration_ms"] // s["calls"]) if s["calls"] else 0
        out_personas.append({"persona": persona, "calls": s["calls"], "errors": s["errors"],
                              "avg_ms": avg_ms})
    return {"by_day": out_days, "by_persona": out_personas, "total_calls": total_calls,
            "exists": True, "today": cutoff_date}


def overview() -> dict:
    """Top-level numbers for the page header."""
    students = list_students()
    n_students = len(students)
    n_open = sum(s.get("open_predictions", 0) for s in students)
    n_resolved = sum(s.get("total_resolved", 0) for s in students)
    n_correct = sum(s.get("total_correct", 0) for s in students)
    classroom_hit = (n_correct / n_resolved) if n_resolved else None
    # Brier across all resolved (approximate, uses each student's lifetime
    # Brier weighted by their resolved count)
    weighted_brier = 0.0
    brier_n = 0
    for s in students:
        b = s.get("lifetime_brier")
        r = s.get("total_resolved") or 0
        if b is not None and r > 0:
            weighted_brier += float(b) * r
            brier_n += r
    overall_brier = (weighted_brier / brier_n) if brier_n else None
    # Today's predictions
    today = datetime.now(timezone.utc).strftime("%Y-%m-%d")
    today_path = DAILY_DIR / today / "predictions.jsonl"
    today_count = 0
    if today_path.exists():
        for _ in _read_jsonl(today_path):
            today_count += 1
    # Recent days that have predictions
    days = []
    if DAILY_DIR.exists():
        for d in sorted(DAILY_DIR.iterdir(), reverse=True):
            if not d.is_dir():
                continue
            pred_path = d / "predictions.jsonl"
            n = 0
            if pred_path.exists():
                n = len(_read_jsonl(pred_path))
            if n:
                days.append({"date": d.name, "n": n})
            if len(days) >= 14:
                break
    return {
        "classroom_root": str(CLASSROOM_ROOT),
        "n_students": n_students,
        "n_open_predictions": n_open,
        "n_resolved": n_resolved,
        "n_correct": n_correct,
        "classroom_hit_rate": classroom_hit,
        "overall_brier": overall_brier,
        "today_date": today,
        "today_predictions": today_count,
        "recent_days": days,
    }


def projections() -> dict:
    """Load the per-student projection bundle written by project_scores.py.
    Filter to live roster + sort by trajectory delta."""
    if not PROJECTIONS_PATH.exists():
        return {"exists": False, "students": [], "classroom": {}, "generated_at": None}
    try:
        data = json.loads(PROJECTIONS_PATH.read_text())
    except json.JSONDecodeError:
        return {"exists": False, "students": [], "classroom": {}, "generated_at": None}
    live = _roster_names()
    rows = [r for r in data.get("students", []) if not live or r.get("name") in live]
    # Sort: smallest projected Brier first (best forecasters)
    rows.sort(key=lambda r: (r.get("projected_brier") is None, r.get("projected_brier") or 1.0))
    return {
        "exists": True,
        "generated_at": data.get("generated_at"),
        "classroom": data.get("classroom", {}),
        "students": rows,
    }


def learning_lab() -> dict:
    """Step-back synthesis for the Learning Lab panel: who's improving, who's
    plateauing, who should be retired."""
    p = projections()
    if not p["exists"]:
        return {"exists": False}
    students = p["students"]
    # Improving = positive trajectory_delta (observed worse than projected)
    improving = [s for s in students if (s.get("trajectory_delta") or 0) > 0.02]
    improving.sort(key=lambda s: -(s.get("trajectory_delta") or 0))
    # Deteriorating = negative trajectory (projected worse than observed)
    deteriorating = [s for s in students if (s.get("trajectory_delta") or 0) < -0.02]
    deteriorating.sort(key=lambda s: (s.get("trajectory_delta") or 0))
    # Most-learned = highest learning_gain (vs naive 70%)
    most_learned = [s for s in students if (s.get("learning_gain") or 0) > 0]
    most_learned.sort(key=lambda s: -(s.get("learning_gain") or 0))
    # Candidates for full-student retirement: all their active setups are bad
    # i.e., zero setups firing recently AND many retired setups.
    full_retire = [s for s in students
                   if s.get("n_setups_retired", 0) >= 2
                   and s.get("n_setups_active_recent", 0) == 0]
    return {
        "exists": True,
        "generated_at": p["generated_at"],
        "classroom": p["classroom"],
        "improving": improving[:10],
        "deteriorating": deteriorating[:10],
        "most_learned": most_learned[:10],
        "retirement_candidates": full_retire[:20],
    }


def latest_predictions(limit: int = 50) -> list[dict]:
    """Most recent predictions across all students."""
    out: list[dict] = []
    if not DAILY_DIR.exists():
        return out
    # Walk recent days until we have enough
    for d in sorted(DAILY_DIR.iterdir(), reverse=True):
        if not d.is_dir():
            continue
        pred_path = d / "predictions.jsonl"
        if not pred_path.exists():
            continue
        rows = _read_jsonl(pred_path)
        out.extend(rows)
        if len(out) >= limit * 2:
            break
    out.sort(key=lambda p: p.get("created_at", ""), reverse=True)
    return out[:limit]
